File: src/modules/AlgoDatCourse/implemented/java/Selectionsort.java
/**
 * Eingabe:    Ein unsortiertes Array A von Zahlen.
 * Ausgabe:    Array A aufsteigend sortiert (in-place).
 */
import java.util.Arrays;

public class Selectionsort {

    /** HAUPTMETHODEN */

    public void selectionsort(int[] A) {

        for (int i = 0; i < A.length - 1; i++) {

            int min = i;

            for (int j = i + 1; j < A.length; j++) {
                if (A[j] < A[min]) {
                    min = j;
                }
            }

            int temp = A[i];
            A[i] = A[min];
            A[min] = temp;
        }
    }

    public static void main(String[] args) {
        int[] A = {3, 1, 4, 1, 5, 9, 2, 6};
        Selectionsort s = new Selectionsort();
        s.selectionsort(A);
        System.out.println(Arrays.toString(A));
    }
}

File: src/modules/AlgoDatCourse/implemented/prolog/Selectionsort.pl
/*
 * Eingabe:    Eine unsortierte Liste von Zahlen.
 * Ausgabe:    Eine aufsteigend sortierte Liste.
 */

:- use_module(library(lists)).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% SELECTIONSORT
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

selectionsort([], []).

selectionsort(List, [Min | Sorted]) :-

    min_list(List, Min),
    delete_first(Min, List, Rest),
    selectionsort(Rest, Sorted).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% HILFSFUNKTIONEN
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

delete_first(X, [X | T], T) :- !.

delete_first(X, [H | T], [H | R]) :-
    delete_first(X, T, R).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% MAIN
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

main :-
    A = [3, 1, 4, 1, 5, 9, 2, 6],
    selectionsort(A, Sorted),
    write('Sorted: '),
    writeln(Sorted).
